### PR TITLE
Remove most string allocations when reading file line by line

### PR DESCRIPTION
**Pull request by Luthaf**
This is a follow-up of #267. With both PR, reading line-oriented formats is roughly 2x faster 🎉 

**Before**
```
$ hyperfine --warmup 3 ./xyz ./lammps-data
Benchmark #1: ./xyz
  Time (mean ± σ):      72.1 ms ±   2.0 ms    [User: 64.2 ms, System: 4.9 ms]
  Range (min … max):    69.5 ms …  77.3 ms    40 runs

Benchmark #2: ./lammps-data
  Time (mean ± σ):      92.0 ms ±   2.7 ms    [User: 82.9 ms, System: 5.6 ms]
  Range (min … max):    87.2 ms …  98.2 ms    31 runs
```

**PR #267**

```
$ hyperfine --warmup 3 ./xyz ./lammps-data
Benchmark #1: ./xyz
  Time (mean ± σ):      56.2 ms ±   1.5 ms    [User: 48.5 ms, System: 4.6 ms]
  Range (min … max):    53.4 ms …  60.4 ms    51 runs

Benchmark #2: ./lammps-data
  Time (mean ± σ):      74.8 ms ±   1.9 ms    [User: 65.8 ms, System: 5.2 ms]
  Range (min … max):    72.0 ms …  79.2 ms    38 runs
```

**Now**

```
$ hyperfine --warmup 3 ./xyz ./lammps-data
Benchmark #1: ./xyz
  Time (mean ± σ):      26.4 ms ±   1.2 ms    [User: 22.0 ms, System: 3.0 ms]
  Range (min … max):    24.7 ms …  29.8 ms    103 runs

Benchmark #2: ./lammps-data
  Time (mean ± σ):      51.7 ms ±   2.8 ms    [User: 45.2 ms, System: 4.3 ms]
  Range (min … max):    47.8 ms …  60.5 ms    58 runs
```

------

The initial idea is from @richardjgowers: instead of allocating a new string for each line, use buffering to read a big block from the file, and return string_view inside this buffer for the line.

If a  line is too long to fit in the buffer (8192 characters by default), the buffer is automatically grown to fit bigger lines.

### CODE REVIEW

**Comment by Luthaf** [comment]
Finally manage to make all compilers happy \o/

**Comment by frodofine** [comment]
This is a large PR, I'll try to get to it when I can.

**Comment by Luthaf** [comment]
@frodofine, do you think you will have time to review this in the next days? If you don't, that's completely fine, I'll do a review round myself :smiley:

**Comment by frodofine** [comment]
Probably not! Sorry!

**Comment by Luthaf** [comment]
No problem, don't worry!

File: include/chemfiles/files/GzFile.hpp [hunk 1]
```diff
@@ -4,58 +4,36 @@
...
+    /// corresponding error message or `nullptr` if no error occured.
+    const char* check_error() const;
 
-private:
-    gzstreambuf buffer_;
+    gzFile file_ = nullptr;
 };
 
-} // namespace chemfiles
+}
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply include/chemfiles/files/GzFile.hpp]: no specific reason =)

I'll look into clang-format, since it can auto-generate these.

**Comment by Luthaf** [comment]
I added a few improvements if you want to look at them @ezavod. Else I'll merge this tomorrow =)

**Pull request merged**